DR2-2318 Changes for DRI migration

MancunianSam opened the pull request:
* Add a filePath parameter. This comes from TDR and from the DRI migration script. This goes into the asset opex
* Allow multiple checksums, either as SHA256ServerSideChecksum from TDR or from a set of checksum_<algorithm> keys from DRI
* Rename the package builder classes and variables to remove tdr. I'll change the package names later.
* Change the package builder so it isn't streaming the metadata files. It's not necessary as they're never that large and it's confusing.
* Add an optional `driBatchReference` parameter which comes from the DRI migration. If it's populated, it's added as an identifier.
* Pass in source system as an environment variable to the package builder
* Fiddled the build.sbt so that it will create two jars with different names for the package builder and aggregator.

techncl commented on .github/workflows/build.yml:
```diff
@@ -31,6 +31,7 @@ jobs:
           (cd ./python/lambdas/copy-files-from-tdr/ && bash deploy.sh)
           zip -jA copy-files-from-tdr python/lambdas/copy-files-from-tdr/lambda_function.py
           cp copy-files-from-tdr target/outputs
+          cp copy-files-from-tdr target/outputs/copy-files-from-dri
```
How come you are copying this to two places?

techncl commented on scala/lambdas/utils/src/main/scala/uk/gov/nationalarchives/utils/ExternalUtils.scala:
```diff
@@ -203,12 +207,28 @@ object ExternalUtils {
+        val algorithm = key.replace(checksumPrefix, "")
+        c.downField(key).as[String].map(fingerprint => Checksum(algorithm, fingerprint))
+      }
+    else
```
Could you not just do

```
    if keys.contains("SHA256ServerSideChecksum") then
      c.downField("SHA256ServerSideChecksum").as[String].map { sha256Checksum =>
        List(Checksum("sha256", sha256Checksum))
      }
    else
      checksumKeys.traverse { key =>
        val algorithm = key.replace(checksumPrefix, "")
        c.downField(key).as[String].map(fingerprint => Checksum(algorithm, fingerprint))
      }

```?

techncl reviewed (commented):
Looks good! Just some minor comments. Also should we add tests for `create_ingest_metadata.py`?

techncl commented on python/dri-migration/create_ingest_metadata.py:
```diff
@@ -0,0 +1,109 @@
+    for name in directory_list:
+        match = pattern.search(name)
+        if match:
+            puid = match.group(1).replace('fmt-', 'fmt/')
```
Do you think it's worth doing this `.replace(f'{prefix}-', f'{prefix}/')` instead?

techncl commented on python/dri-migration/create_ingest_metadata.py:
```diff
@@ -0,0 +1,109 @@
+
+account_number = os.environ["ACCOUNT_NUMBER"]
+environment = os.environ["ENVIRONMENT"]
+bucket = f"{environment}-dr2-ingest-raw-cache"
```
Should this be an environment variable?

techncl commented on scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/TestUtils.scala:
```diff
@@ -78,12 +95,14 @@ object TestUtils:
           for {
             jsonString <- publisher.toStreamBuffered[IO](1024).map(_.array().map(_.toChar).mkString).compile.string
             json <- IO.fromEither(decode[List[MetadataObject]](jsonString))
+            _ <- IO.println(jsonString)
```
Is this supposed to be here?

techncl commented on python/dri-migration/create_ingest_metadata.py:
```diff
@@ -0,0 +1,109 @@
+        file_path = row[key_indexes["FILE_PATH"]]
+        full_path = row[key_indexes["FULLPATH"]]
+        checksum_data = json.loads(row[key_indexes["FIXITIES"]])
+        metadata = {
```
Should we use the metadata schema to validate that we have all of the necessary fields?

techncl commented:
Could you add a README for the DRI migration project?

MancunianSam commented on build.sbt:
```diff
@@ -1,11 +1,12 @@
 import Dependencies.*
-ThisBuild / scalaVersion := "3.7.1"
+ThisBuild / scalaVersion := "3.7.0"
```
I was fiddling with the code analysis tools and it didn't have a version for 3.7.1 I'm not doing that now so I've put it back.

MancunianSam commented on python/dri-migration/create_ingest_metadata.py:
```diff
@@ -0,0 +1,106 @@
+def create_skeleton_suite_lookup(prefix):
+    path = f"{os.environ['DROID_PATH']}\\{prefix}"
+
+    pattern = re.compile(r'((x-)?fmt-\d{1,5})-signature-id-\d{1,5}(\.[A-Za-z]{1,10})')
```
1. This isn't a long regex. I did try it but I don't think it helped readability that much. 
2. I tried this and it doesn't seem to work, it doesn't find any matches like this.

MancunianSam commented on python/dri-migration/create_ingest_metadata.py:
```diff
@@ -0,0 +1,106 @@
+page_size = 100
+
+s3_client = boto3.client("s3")
+sqs_client = boto3.client("s3")
```
Yeah it should, I've changed it.

MancunianSam commented on scala/lambdas/build.sbt:
```diff
@@ -267,33 +284,56 @@ lazy val postIngestStateChangeHandler = (project in file("postingest-state-chang
   .dependsOn(utils, dynamoFormatters)
+  .settings(packageBuilderSettings)
+
+lazy val preingestDriPackageBuilder = (project in file("preingest-tdr-package-builder"))
```
Yeah, using the same source code but for a different module. I'll rename the directory to `preingest-package-builder` at some point but not now.

MancunianSam commented on scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json:
```diff
@@ -18,8 +18,7 @@
     "transferringBody": {
-      "type": "string",
-      "minLength": 1
+      "type": "string"
```
No, that's not right. I think I thought the min length was making it mandatory but that's only if populated. I've put the minLength back.

techncl replied:
Ah, I see. Could we add a test that checks this?

MancunianSam replied:
I've just tried this with the minLength and no transferringBody key and it fails which we don't want it to so I've removed the minLength again.

techncl replied:
That's weird. Wait, is this happening because you've added `packageMetadata.transferringBody.getOrElse("")`? i.e. even if the transferring body is missing we'll still make it an empty string?

MancunianSam replied:
Oh yeah, that'll be it. I was being lazy, I'll make the asset field optional.

MancunianSam replied:
I've modified one of the validator tests to remove the transferring body field and added another test to check we get an error if it's there but empty and this seems to work.

MancunianSam commented:
I've added a README. 

I've also made the source system an enum and changed the case class fields for `PackageMetadata` so the case of the fields is consistent now. It's all in this commit https://github.com/nationalarchives/dr2-ingest/pull/586/commits/12c650d06f3a76c1dc2f4cbc9d56958b8726036d

techncl commented on scala/lambdas/preingest-tdr-package-builder/README.md:
```diff
@@ -52,3 +54,4 @@ The lambda outputs this json.
 | LOCK_DDB_TABLE_GROUPID_GSI_NAME | The lock table global secondary index name in Dynamo                                                                                    |
 | OUTPUT_BUCKET_NAME              | The bucket the TDR file and metadata file are stored in                                                                                 |
 | CONCURRENCY                     | An optional parameter which configures how many items the lambda processes in parallel. Used for performance tweaking. Defaults to 1000 |
+| SOURCE_SYSTEM                   | TDR or DRI                                                                                                                              |
```
This is missing the TRE source system

MancunianSam commented:
I've added a definitely not copilot generated python test for the migration script. It only does the happy path but that's fine, I don't want to get bogged down checking for errors.

MancunianSam commented on python/dri-migration/create_ingest_metadata_test.py:
```diff
@@ -0,0 +1,70 @@
+        os.environ['ENVIRONMENT'] = 'testenv'
+
+        mock_create_skeleton.return_value = {
+            "fmt/123": {"file_path": "/test/file1"}
```
Yeah ok. I think the easiest way is to have to rows, one which returns fmt/123 and one for x-fmt/123. I've changed the ids slightly on the rows to test that we're getting the right data from multiple rows. 

I've also added in a test for `create_skeleton_suite_lookup` and `calculate_checksum` and these do check the slightly unhappy path.
